packages/ui/react-ui-editor/src/components/Toolbar/Toolbar.stories.tsx

```diff
@@ -50,11 +61,12 @@ const Story: FC<{ content: string }> = ({ content }) => {
     return null;
   }
 
+  // TODO Wire up the toolbar state to the editor properly
   return (
     <div className={mx(fixedInsetFlexLayout, groupSurface)}>
       <div className='flex h-full justify-center'>
         <div className='flex flex-col h-full w-[800px]'>
-          <Toolbar.Root onAction={handleAction}>
+          <Toolbar.Root onAction={handleAction} state={editorView ? getFormatting(editorView.state) : emptyFormatting}>
```

Comment:
i don't see how/when this is updated since we don't want to re-render Toolbar on each document update, so getFormatting will be frozen.

i think i would create a hook to decouple the dependencies (invert the dependency) to be used by the external component that renders the toolbar.

```tsx
const useToolbarState = () => {
  const [state, setState] = useState({});
  const observer = useMemo(() => EditorView.updateListener.of((update) => {
    setState(getFormatting(update.view));
  }, []);
  return [
    state,
    observer
  ];
}

const Story = () => {
  const [state, observer] = useToolbarState();

  return (
    <Toolbar state={state} />
    <TextEditor ref={ref} extensions={[observer]} />
  );
}
```

Reply:
Thanks, that's what I was looking for.

Reply:
NOTE: you'd want to not create a new state object if the state hasn't changed since i know updateListener is called very frequently.